neuralop/layers/codano.py

```diff
@@ -0,0 +1,607 @@
+
+        mixer_args = dict(
+            n_modes=n_modes,
+            output_scaling_factor=1,
+            non_linearity=non_linearity,
+            norm='instance_norm',
+            fno_skip=fno_skip,
+            conv_module=SpectralConvolution,
+        )
+        # We have an option to make the last operator (MLP in regular
+        # Transformer block) permutation equivariant. i.e., applying the
+        # operator per variable or applying the operator on the whole channel
+        # (like regular FNO).
+        if permutation_eq:
+            self.mixer = FNOBlocks(
+                in_channels=self.mixer_token_codimension,
+                out_channels=self.mixer_token_codimension,
+                apply_skip=True,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(self.token_codimension)
+            self.norm2 = Normalizer(self.mixer_token_codimension)
+            self.mixer_out_normalizer = Normalizer(
+                self.mixer_token_codimension)
+
+        else:
+            self.mixer = FNOBlocks(
+                in_channels=codimension_size,
+                out_channels=codimension_size,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(codimension_size)
+            self.norm2 = Normalizer(codimension_size)
+            self.mixer_out_normalizer = Normalizer(codimension_size)
+
+    def forward(self, *args):
+        raise NotImplementedError(
+            "Use a proper subclass of TNOBlock (i.e. TnoBlock2d or TNOBlock3D).")
+
+
+class CODABlocks2D(CODABlocks):
+    def __init__(self, *args, **kwargs):
+        Normalizer = kwargs.get("Normalizer")
+        if Normalizer is None:
+            Normalizer = partial(nn.InstanceNorm2d, affine=True)
+        kwargs["Normalizer"] = Normalizer
+
+        Convolution = kwargs.get("SpectralConvolution")
+        if Convolution is None:
+            Convolution = SpectralConv2d
+        kwargs["SpectralConvolution"] = Convolution
+
+        super().__init__(*args, **kwargs)
+
+    # XXX rewrite comments on TNO*3D
+    def compute_attention(self, xa, batch_size):
+        """Compute the key-query-value variant of the attention matrix.
+
+        Assumes input ``xa`` has been normalized.
+        """
+        k = self.K(xa)
+        q = self.Q(xa)
+        v = self.V(xa)
+
+        value_x, value_y = v.shape[-2], v.shape[-1]
+
+        t = k.size(0) // batch_size  # Compute the time dimension `t`
+        d = k.size(1) // self.n_head  # Compute the dimension `d`
+
+        k = k.view(batch_size, t, self.n_head, d, k.size(-2),
+                   k.size(-1))
+        q = q.view(batch_size, t, self.n_head, d, q.size(-2),
+                   q.size(-1))
+        v = v.view(batch_size, t, self.n_head, d, v.size(-2),
+                   v.size(-1))
+        k = k.permute(0, 2, 1, 3, 4, 5).contiguous()
```

Comment:
These permutations make me think this code isn't general for all spatial dimensions. Could you either be more clear about which shapes/dims this is valid for, or change the reshaping/permuting logic to handle all dimensions?

Reply:
Assertion added.
The rearrange operation is much clearer using eiops rearrange operation.
I had that in the first place. But it adds a new dependency. But it produces much readable code. As @JeanKossaifi  suggested, I removed that.

Reply:
Two ideas:

1. Combine CODA 2d and 3D into one layer as I suggested, and make this logic more general (check the reshaping/permuting in `neuralop/models/gino.py`)
2. Could you use `torch.einsum`